typhos/alarm.py

```diff
@@ -0,0 +1,380 @@
+                self._channels.clear()
+                for channel in self._channels:
+                    if hasattr(channel, 'disconnect'):
+                        channel.disconnect()
+            # Load new channel
+            self._channel = str(value)
+            channel = HappiChannel(
+                address=self._channel,
+                tx_slot=self._tx,
+                connection_slot=partial(self.update_connection,
+                                        addr=self._channel),
+                severity_slot=partial(self.update_severity,
+                                      addr=self._channel),
+                )
+            self._channels = [channel]
+            # Connect the channel to the HappiPlugin
+            if hasattr(channel, 'connect'):
+                channel.connect()
+
+    def _tx(self, value):
+        """Receive information from happi channel"""
+        self.add_device(value['obj'])
+
+    def init_alarm_state(self):
+        self.addr_connected = {}
+        self.addr_severity = {}
+        self.addr_channels = {}
+        self.device_channels = {}
+        self.alarm_summary = AlarmLevel.DISCONNECTED
+        self.set_alarm_color(AlarmLevel.DISCONNECTED)
+
+    def channels(self):
+        """
+        Let pydm know about our pydm channels.
+        """
+        ch = list(self._channels)
+        for lst in self.device_channels.values():
+            ch.extend(lst)
+        return ch
+
+    def add_device(self, device):
+        """
+        Initialize our alarm handling when adding a device.
+        """
+        super().add_device(device)
+        self.setup_alarm_config(device)
+
+    def clear_all_alarm_configs(self):
+        """
+        Reset this widget down to the "no alarm handling" state.
+        """
+        channels = self.addr_channels.values()
+        for ch in channels:
+            ch.disconnect()
+        self.init_alarm_state()
+
+    def setup_alarm_config(self, device):
+        """
+        Add a device to the alarm summary.
+
+        This will pick PVs based on the device kind and the configured kind
+        level, configuring the PyDMChannels to update our alarm state and
+        color when we get updates from our PVs.
+        """
+        sigs = get_all_signals_from_device(
+            device,
+            filter_by=KIND_FILTERS[self._kind_level]
+            )
+        channel_addrs = [channel_from_signal(sig) for sig in sigs]
+        for sig in sigs:
+            if not isinstance(sig, EpicsSignalBase):
+                register_signal(sig)
+        channels = [
+            PyDMChannel(
+                address=addr,
+                connection_slot=partial(self.update_connection, addr=addr),
+                severity_slot=partial(self.update_severity, addr=addr),
+                )
+            for addr in channel_addrs]
+
+        self.device_channels[device.name] = channels
+        for ch in channels:
+            self.addr_channels[ch.address] = ch
```

Comment:
I think there's arguments for/against the separate dictionaries as show here, but just as a refactor consideration:

`addr_info[ch.address] = {"channel": ..., "connected": ..., "severity": ...}` - or using a dataclass to hold that information instead of a dict, even - could be an alternative.

Reply:
Right, that's a good point. I'll think about this and probably refactor a bit

Reply:
After playing with this a bit, dataclasses are the way to go. I'll probably never use a nested dict ever again.